source/sigv4.c

```diff
@@ -794,6 +918,308 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+                {
+                    *pCurrBufLoc = tolower( pData[ index ] );
+                }
+
+                pCurrBufLoc++;
+                numOfBytesCopied++;
+                buffRemaining -= 1;
+            }
+        }
+
+        /* Add the ending separating character passed to the function.
+         * Note: Space for the separator character is accounted for while copying
+         * header field data to canonical request buffer. */
+        if( status == SigV4Success )
+        {
+            assert( buffRemaining >= 1 );
+            *pCurrBufLoc = separator;
+            pCurrBufLoc++;
+            numOfBytesCopied++;
+            canonicalRequest->pBufCur = pCurrBufLoc;
+            canonicalRequest->bufRemaining = ( buffRemaining - 1 );
+        }
+
+        return status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendSignedHeaders( size_t headerCount,
+                                              uint32_t flags,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t headerIndex = 0, keyLen = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+        const char * headerKey;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+        assert( headerCount > 0 );
+
+        for( headerIndex = 0; headerIndex < headerCount; headerIndex++ )
+        {
+            assert( ( canonicalRequest->pHeadersLoc[ headerIndex ].key.pData ) != NULL );
+            keyLen = canonicalRequest->pHeadersLoc[ headerIndex ].key.dataLen;
+
+            headerKey = canonicalRequest->pHeadersLoc[ headerIndex ].key.pData;
+
+            sigV4Status = copyHeaderStringToCanonicalBuffer( headerKey, keyLen, flags, ';', canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            /* Replacing the last ';' with '\n' as last header does need to have ';'. */
+            *( canonicalRequest->pBufCur - 1 ) = '\n';
+        }
+
+        return sigV4Status;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static SigV4Status_t appendCanonicalizedHeaders( size_t headerCount,
+                                                     uint32_t flags,
+                                                     CanonicalContext_t * canonicalRequest )
+    {
+        size_t headerIndex = 0, keyLen = 0, valLen = 0;
+        const char * value;
+        const char * headerKey;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+        assert( headerCount > 0 );
+
+        for( headerIndex = 0; headerIndex < headerCount; headerIndex++ )
+        {
+            assert( canonicalRequest->pHeadersLoc[ headerIndex ].key.pData != NULL );
```

Comment:
Should there be an assert for the header value OR is it fine for the header value to be empty?

Reply:
According to here:
https://stackoverflow.com/questions/12130910/how-to-interpret-empty-http-accept-header

Looks like it's possible for the value to empty. Although it'd be pointless to have an empty header value if its value is meant to be signed for the purposes of SigV4.

Reply:
I checked the coreHTTP library so when we add HTTP header, coreHTTP library does allows empty keys or values.

Reply:
In that case, let's add a comment in the API documentation stating that empty header values (or header values containing whitespaces) are not supported